### PR TITLE
Adapt export API for events (including all detail levels)

### PR DESCRIPTION
**Pull request by nop33**
### Changes from legacy:
* `id` of chairpersons, speakers, authors, coauthors, conveners:
    * was a friendly id. After this PR it will change to the `EventPerson.id`.
    * was a `str`, now it's an `int`.
* `id` of the event was a `str` now it's an `int`
* session block id was in the format `<session friendly id>-<block friendly id>`, now it's the postgres `SessionBlock.id`.
* There is no `description` in `SessionBlock` anymore.

### Dependencies
- [x] Un-comment session `startDate` and `endDate` when https://github.com/indico/indico/pull/2272 gets merged

### CODE REVIEW

File: indico/modules/events/api.py [hunk 1]
```diff
@@ -18,16 +18,22 @@
 import itertools
 import re
 import pytz
+from hashlib import md5
```
Comment (ThiefMaster) [inline indico/modules/events/api.py]: h>d

File: indico/modules/events/api.py [hunk 2]
```diff
@@ -214,11 +220,24 @@ def _addOccurrences(self, fossil, obj, startDT, endDT):
...
 
     def _makeFossil(self, obj, iface):
+        legacy_obj = obj.as_legacy if isinstance(obj, Event) else obj
         return fossilize(obj, iface, tz=self._tz, naiveTZ=self._serverTZ,
                          filters={'access': self._userAccessFilter},
-                         canModify=obj.canModify(self._aw),
+                         canModify=legacy_obj.canModify(self._aw),
                          mapClassType={'AcceptedContribution': 'Contribution'})
 
+    def _get_query_options(self, include_contr=False):
```
Comment (ThiefMaster) [inline indico/modules/events/api.py]: `include_contribs`

File: indico/modules/events/api.py [hunk 3]
```diff
@@ -235,19 +254,19 @@ def filter(obj):
...
                          Event.category_chain.overlap(map(int, idlist)),
                          Event.happens_between(self._fromDT, self._toDT))
-                 .options(acl_user_strategy))
+                 .options(*self._get_query_options()))
         return self._process((x.as_legacy for x in query), filter)
 
     def event(self, idlist):
+        events = (Event.find(Event.id.in_(idlist),
+                             ~Event.is_deleted)
+                       .options(*self._get_query_options(include_contr=True))
```
Comment (ThiefMaster) [inline indico/modules/events/api.py]: that's now how we align multiline queries ;)

```python
events = (Event.find(Event.id.in_(idlist), ~Event.is_deleted)
          .options(*self._get_query_options(include_contribs=True))
          .all())
```

File: indico/modules/events/api.py [hunk 4]
```diff
@@ -255,8 +274,101 @@ def _iterate_objs(objIds):
                 obj = ch.getById(objId, True)
                 if obj is not None:
                     yield obj
+        return self.serialize_events(events)
+        # return self._process(_iterate_objs(idlist))
+
+    def serialize_events(self, events):
+        results = []
```
Comment (ThiefMaster) [inline indico/modules/events/api.py]: return map(self._build_event_api_data, events)

File: indico/modules/events/contributions/models/contributions.py [hunk 5]
```diff
@@ -295,6 +296,14 @@ def speakers(self):
         return [person_link for person_link in self.person_links if person_link.is_speaker]
 
     @property
+    def primary_authors(self):
```
Comment (ThiefMaster) [inline indico/modules/events/contributions/models/contributions.py]: @mic4ael added properties for those.. maybe cherrypick his commit?

f92dddd11f2176021f25dde1bc0bb5d20bc7a7b7

File: indico/modules/events/api.py [hunk 6]
```diff
@@ -255,8 +273,141 @@ def _iterate_objs(objIds):
...
+    def _build_contribution_api_data(self, contrib, include_subcontrs=False):
+        data = {
+            '_type': 'Contribution',
+            '_fossil': self.fossils_mapping['contribution'].get(self._detail_level, None),
+            'id': contrib.id,
+            'title': contrib.title,
+            'startDate': self._serialize_date(contrib.start_dt) if contrib.start_dt else None,
+            'endDate': self._serialize_date(contrib.start_dt + contrib.duration) if contrib.start_dt else None,
+            'duration': contrib.duration.seconds / 60,
+            'roomFullname': contrib.room_name,
```
Comment (ThiefMaster) [inline indico/modules/events/api.py]: was it really `roomFullname` and not `roomFullName`?

File: indico/modules/events/api.py [hunk 7]
```diff
@@ -255,8 +273,141 @@ def _iterate_objs(objIds):
...
+
+    def _build_contribution_api_data(self, contrib, include_subcontrs=False):
+        data = {
+            '_type': 'Contribution',
+            '_fossil': self.fossils_mapping['contribution'].get(self._detail_level, None),
+            'id': contrib.id,
+            'title': contrib.title,
+            'startDate': self._serialize_date(contrib.start_dt) if contrib.start_dt else None,
+            'endDate': self._serialize_date(contrib.start_dt + contrib.duration) if contrib.start_dt else None,
+            'duration': contrib.duration.seconds / 60,
```
Comment (ThiefMaster) [inline indico/modules/events/api.py]: `contrib.duration.total_seconds() // 60`

File: indico/modules/events/api.py (hunk 6 above)
Comment (nop33) [inline indico/modules/events/api.py]: image

File: indico/modules/events/api.py [hunk 8]
```diff
@@ -255,8 +273,141 @@ def _iterate_objs(objIds):
...
+    def _serialize_date(self, date):
+        if date:
+            return {
+                'date': str(date.date()),
+                'time': str(date.time()),
+                'tz': str(date.tzinfo)
+            }
+
+    def _serialize_persons(self, persons, person_type):
+        result = []
```
Comment (ThiefMaster) [inline indico/modules/events/api.py]: use a list comprehension!

File: indico/modules/events/api.py [hunk 9]
```diff
@@ -255,8 +273,141 @@ def _iterate_objs(objIds):
...
+        for person in persons:
+            result.append(self._serialize_person(person, person_type))
+        return result
+
+    def _serialize_person(self, person, person_type):
```
Comment (ThiefMaster) [inline indico/modules/events/api.py]: You sometimes pass `User` objects here and sometimes person links. person links do not have an email (since that's only stored on the person itself) - you either need to add an `email` property to `PersonLinkBase` or get the email from the actual person instead of the person link. the property is probably cleaner...

File: indico/modules/events/api.py [hunk 10]
```diff
@@ -214,11 +220,24 @@ def _addOccurrences(self, fossil, obj, startDT, endDT):
...
 
+    def _get_query_options(self, include_contribs=False):
+        acl_user_strategy = joinedload('acl_entries').joinedload('user')
+        # remote group membership checks will trigger a load on _all_emails
+        # but not all events use this so there's no need to eager-load them
+        # acl_user_strategy.noload('_primary_email')
+        # acl_user_strategy.noload('_affiliation')
+        creator_strategy = joinedload('creator')
+        contributions_strategy = subqueryload('contributions')
+        if include_contribs:
```
Comment (ThiefMaster) [inline indico/modules/events/api.py]: what about sessions? check how many queries you get when exporting a category with multiple events on the highest detail level

File: indico/modules/events/api.py [hunk 11]
```diff
@@ -235,19 +254,18 @@ def filter(obj):
...
                  .filter(~Event.is_deleted,
                          Event.category_chain.overlap(map(int, idlist)),
                          Event.happens_between(self._fromDT, self._toDT))
-                 .options(acl_user_strategy))
+                 .options(*self._get_query_options()))
         return self._process((x.as_legacy for x in query), filter)
 
     def event(self, idlist):
+        events = (Event.find(Event.id.in_(idlist), ~Event.is_deleted)
+                  .options(*self._get_query_options(include_contribs=True))
```
Comment (ThiefMaster) [inline indico/modules/events/api.py]: what if the detail level doesn't include contributions?

File: indico/modules/events/api.py [hunk 12]
```diff
@@ -255,8 +273,243 @@ def _iterate_objs(objIds):
...
+        }
+
+    def _serialize_session(self, session):
+        return {
+            'folders': build_folders_api_data(session),
+            # TODO: Add when #2272 gets merged
+            # 'startDate': self._serialize_date(session.start_dt) if session.blocks else None,
+            # 'endDate': self._serialize_date(session.end_dt) if session.blocks else None,
+            '_type': 'Session',
+            'sessionConveners': [self._serialize_convener(convener) for convener in session.conveners],
```
Comment (ThiefMaster) [inline indico/modules/events/api.py]: `map(self._serialize_convener, session.conveners)`

File: indico/modules/events/api.py [hunk 13]
```diff
@@ -255,8 +273,243 @@ def _iterate_objs(objIds):
...
+            'title': session.title,
+            'color': '#{}'.format(session.colors.background),
+            'textColor': '#{}'.format(session.colors.text),
+            'description': session.description,
+            'material': build_material_legacy_api_data(session),
+            'isPoster': session.is_poster,
+            'url': url_for('event.sessionDisplay', confId=session.event_id, sessionId=session.friendly_id,
+                           _external=True),  # TODO: Change to new endpoint once it's there
+            'protectionURL': url_for('sessions.session_protection', session.event_new, session, _external=True),
```
Comment (ThiefMaster) [inline indico/modules/events/api.py]: not sure if keeping this URL makes any sense; the page is not meant to be loaded outside a dialog. maybe better to remove it?

Comment (nop33) [inline indico/modules/events/api.py]: With @pferreir 's permission, I'll gladly remove it.

File: indico/modules/events/api.py [hunk 14]
```diff
@@ -255,8 +273,243 @@ def _iterate_objs(objIds):
...
+            'id': event.id,
+            'title': event.title,
+            'description': event.description,
+            'startDate': self._serialize_date(event.start_dt),
+            'timezone': event.timezone,
+            'endDate': self._serialize_date(event.end_dt),
+            'room': event.room_name,
+            'location': event.venue_name,
+            'address': event.address,
+            'type': event.type
```
Comment (ThiefMaster) [inline indico/modules/events/api.py]: did we send `simple_event` or `lecture` here before? `getType()` returns `simple_event` but `type` returns `lecture`

File: indico/modules/events/api.py (hunk 11 above)
Comment (nop33) [inline indico/modules/events/api.py]: hm, you're right. I'll move the line
```python
self._detail_level = get_query_parameter(request.args.to_dict(), ['d', 'detail'])
```
above this.

File: indico/modules/events/api.py [hunk 15]
```diff
@@ -255,8 +273,243 @@ def _iterate_objs(objIds):
...
+        }
+
+    def _build_event_api_data(self, event):
+        self._detail_level = get_query_parameter(request.args.to_dict(), ['d', 'detail'])
```
Comment (ThiefMaster) [inline indico/modules/events/api.py]: i don't think this is a good place to set this. it's the same for each event anyway

File: indico/modules/events/api.py [hunk 16]
```diff
@@ -255,8 +273,243 @@ def _iterate_objs(objIds):
...
+            'roomMapURL': event.room.map_url if event.room else None,
+            'visibility': Conversion.visibility(event.as_legacy),
+            'folders': build_folders_api_data(event),
+            'chairs': self._serialize_persons(event.person_links, person_type='ConferenceChair'),
+            'material': build_material_legacy_api_data(event) + [build_note_legacy_api_data(event.note)]
+        })
+        if self._detail_level in {'contributions', 'subcontributions'}:
+            data['contributions'] = []
+            for contribution in event.contributions:
+                include_subcontrs = self._detail_level in {'subcontributions', 'sessions'}
```
Comment (ThiefMaster) [inline indico/modules/events/api.py]: `include_subcontribs`.. let's not start using two different ways of abbreviating "contribution" ;)

File: indico/modules/events/api.py [hunk 17]
```diff
@@ -255,8 +273,243 @@ def _iterate_objs(objIds):
...
+        elif self._detail_level == 'sessions':
+            # Contributions without a session
+            data['contributions'] = []
+            for contribution in event.contributions:
+                if not contribution.session:
+                    serialized_contrib = self._build_contribution_api_data(contribution)
+                    data['contributions'].append(serialized_contrib)
+
+            data['sessions'] = []
+            for session in event.sessions:
```
Comment (ThiefMaster) [inline indico/modules/events/api.py]: don't name variables `session`; it collides with `flask.session` in case it's used. use `sess` or `session_`

File: indico/modules/events/api.py [hunk 18]
```diff
@@ -255,8 +273,243 @@ def _iterate_objs(objIds):
...
+                data['sessions'].extend(self._build_session_api_data(session))
+        return data
+
+    def _build_session_event_api_data(self, event):
+        data = self._build_event_basic_api_data(event)
+        data.update({
+            '_fossil': 'conference',
+            'adjustedStartDate': self._serialize_date(event.as_legacy.getAdjustedStartDate()),
+            'adjustedEndDate': self._serialize_date(event.as_legacy.getAdjustedEndDate()),
+            'bookedRooms': Conversion.reservationsList(event.reservations.all()),
```
Comment (ThiefMaster) [inline indico/modules/events/api.py]: eventually this should be changed to a non-dynamic relationship so you can subqueryload it instead of triggering a new query here (for each event)

File: indico/modules/events/api.py [hunk 19]
```diff
@@ -255,8 +273,243 @@ def _iterate_objs(objIds):
...
+            'location': contrib.venue_name,
+            'type': contrib.type.name if contrib.type else None,
+            'description': contrib.description,
+            'folders': build_folders_api_data(contrib),
+            'url': url_for('event.contributionDisplay', confId=contrib.event_id, contribId=contrib.friendly_id,
+                           _external=True),
+            'material': build_material_legacy_api_data(contrib),
+            'speakers': self._serialize_persons([x.person for x in contrib.speakers],
+                                                person_type='ContributionParticipation'),
+            'primaryauthors': self._serialize_persons([x.person for x in contrib.primary_authors],
```
Comment (ThiefMaster) [inline indico/modules/events/api.py]: it was written like this?

File: indico/modules/events/sessions/models/blocks.py [hunk 20]
```diff
@@ -127,9 +131,6 @@ def can_manage(self, user, allow_admin=True):
         else:
             return False
 
-    def can_access(self, user, allow_admin=True):
-        return self.session.can_access(user, allow_admin=allow_admin)
```
Comment (ThiefMaster) [inline indico/modules/events/sessions/models/blocks.py]: why did you remove this?

File: indico/modules/events/api.py (hunk 19 above)
Comment (nop33) [inline indico/modules/events/api.py]: image

File: indico/modules/events/sessions/models/blocks.py (hunk 20 above)
Comment (nop33) [inline indico/modules/events/sessions/models/blocks.py]: Check this line ;)
https://github.com/indico/indico/pull/2296/files/22646562c0e83102506fe3cafe31f5bff9c031c6#diff-2852b3896b207ee61528fd8aba9593aeR102

File: indico/modules/events/api.py (hunk 14 above)
Comment (nop33) [inline indico/modules/events/api.py]: As shit, it's `simple_event`... Should I use the `getType()` or just handle the case of lecture specifically?

File: indico/modules/events/api.py (hunk 10 above)
Comment (nop33) [inline indico/modules/events/api.py]: How do you count the queries? In the `indico_dblog` I can scroll to see up to 20 queries but there are more.

File: indico/modules/events/sessions/models/blocks.py (hunk 20 above)
Comment (mic4ael) [inline indico/modules/events/sessions/models/blocks.py]: The link is invalid

Comment (nop33) [inline indico/modules/events/sessions/models/blocks.py]: The property is repeated in line 103.

File: indico/modules/events/api.py (hunk 10 above)
Comment (nop33) [inline indico/modules/events/api.py]: nvm, saw it at the bottom!

Comment (ThiefMaster) [inline indico/modules/events/api.py]: You can configure your terminal to show a bigger scrollback buffer. I usually set it to ~100k lines

File: indico/modules/attachments/api/util.py [hunk 21]
```diff
@@ -27,9 +27,17 @@
 from MaKaC.common.contextManager import ContextManager
 
 
+def get_event(linked_object):
+    from MaKaC.conference import Category, Conference
+
+    if isinstance(linked_object, Conference):
```
Comment (ThiefMaster) [inline indico/modules/attachments/api/util.py]: How do you end up with Conference objects here? I don't think that should happen..

File: indico/modules/events/api.py [hunk 22]
```diff
@@ -235,19 +261,19 @@ def filter(obj):
...
-        # remote group membership checks will trigger a load on _all_emails
-        # but not all events use this so there's no need to eager-load them
-        acl_user_strategy.noload('_primary_email')
-        acl_user_strategy.noload('_affiliation')
         query = (Event.query
                  .filter(~Event.is_deleted,
                          Event.category_chain.overlap(map(int, idlist)),
                          Event.happens_between(self._fromDT, self._toDT))
-                 .options(acl_user_strategy))
+                 .options(*self._get_query_options()))
```
Comment (ThiefMaster) [inline indico/modules/events/api.py]: you need to pass the detail level here too

File: indico/modules/events/api.py [hunk 23]
```diff
@@ -214,11 +220,31 @@ def _addOccurrences(self, fossil, obj, startDT, endDT):
...
 
     def _makeFossil(self, obj, iface):
+        legacy_obj = obj.as_legacy if isinstance(obj, Event) else obj
         return fossilize(obj, iface, tz=self._tz, naiveTZ=self._serverTZ,
                          filters={'access': self._userAccessFilter},
-                         canModify=obj.canModify(self._aw),
+                         canModify=legacy_obj.canModify(self._aw),
                          mapClassType={'AcceptedContribution': 'Contribution'})
 
+    def _get_query_options(self, detail_level=None):
```
Comment (ThiefMaster) [inline indico/modules/events/api.py]: the default value doesn't make much sense here

File: indico/modules/events/api.py [hunk 24]
```diff
@@ -214,11 +220,31 @@ def _addOccurrences(self, fossil, obj, startDT, endDT):
...
+        acl_user_strategy = joinedload('acl_entries').joinedload('user')
+        # remote group membership checks will trigger a load on _all_emails
+        # but not all events use this so there's no need to eager-load them
+        # acl_user_strategy.noload('_primary_email')
+        # acl_user_strategy.noload('_affiliation')
+        creator_strategy = joinedload('creator')
+        contributions_strategy = subqueryload('contributions')
+        subcontributions_strategy = subqueryload('contributions').subqueryload('subcontributions')
+        sessions_strategy = subqueryload('sessions')
+        options = acl_user_strategy, creator_strategy
```
Comment (ThiefMaster) [inline indico/modules/events/api.py]: A list seems cleaner, and then `.append()` the optional ones below

File: indico/modules/events/api.py [hunk 25]
```diff
@@ -235,19 +261,19 @@ def filter(obj):
...
                          Event.category_chain.overlap(map(int, idlist)),
                          Event.happens_between(self._fromDT, self._toDT))
-                 .options(acl_user_strategy))
+                 .options(*self._get_query_options()))
         return self._process((x.as_legacy for x in query), filter)
 
     def event(self, idlist):
+        self._detail_level = get_query_parameter(request.args.to_dict(), ['d', 'detail'])
+        events = (Event.find(Event.id.in_(idlist), ~Event.is_deleted)
+                  .options(*self._get_query_options(detail_level=self._detail_level))
```
Comment (ThiefMaster) [inline indico/modules/events/api.py]: i don't think the kwarg syntax is really needed here

File: indico/modules/events/api.py [hunk 26]
```diff
@@ -255,8 +281,242 @@ def _iterate_objs(objIds):
...
+            'address': convener.address,
+            '_fossil': 'conferenceParticipation',
+            'fullName': convener.full_name,
+            'id': convener.person_id
+        }
+
+    def _serialize_session(self, session_):
+        return {
+            'folders': build_folders_api_data(session_),
+            # TODO: Add when #2272 gets merged
```
Comment (ThiefMaster) [inline indico/modules/events/api.py]: I'd just pick those changes (manually if needed).

File: indico/modules/events/api.py (hunk 23 above)
Comment (nop33) [inline indico/modules/events/api.py]: Why not? The detail level is an optional parameter in the URL, so why not here too?

Comment (ThiefMaster) [inline indico/modules/events/api.py]: ah, I thought we'd have a default value that's not None in case nothing is specified. Anyway, you always pass one here so it's a bit pointless to have a default value in my opinion
